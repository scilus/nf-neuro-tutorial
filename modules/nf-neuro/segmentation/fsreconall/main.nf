process SEGMENTATION_FSRECONALL {
    tag "$meta.id"
    label 'process_single'

    // Note. Freesurfer is already on Docker. See documentation on
    // https://hub.docker.com/r/freesurfer/freesurfer
    container "freesurfer/freesurfer:7.4.1"

    input:
        tuple val(meta), path(anat), path(fs_license) /* optional, value = [] */

    output:
        tuple val(meta), path("*__recon_all")   , emit: recon_all_out_folder
        path "versions.yml"                     , emit: versions

    when:
    task.ext.when == null || task.ext.when

    // Note. In dsl1, we used an additional option:   -parallel -openmp $params.nb_threads.
    // Removed here.
    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def dev_debug_test = task.ext.debug ? task.ext.debug : ""  // If true, we will only run the help (for unit tests )
    """
    # Manage the license. (Save old one if existed.)
    if [[ ! -f "$fs_license" ]]
    then
        echo "License not given in input, or not found. Will probably fail. "
    else
        cp "$fs_license" .license
        here=`pwd`
        export FS_LICENSE=\$here/.license
    fi

    if [ -z $dev_debug_test ]
    then
        # Run the main script
        export SUBJECTS_DIR=`pwd`
        recon-all -i $anat -s ${prefix}__recon_all -all
    else
        # (for developers: unit tests: skip the long processing. help only.)
        export SUBJECTS_DIR=`pwd`
        recon-all -i $anat -s ${prefix}__recon_all -autorecon1 -dontrun
    fi

    # Remove the license
    if [ ! $fs_license = [] ]; then
        rm .license
    fi


    # Finish
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        freesurfer: \$(mri_convert -version | grep "freesurfer" | sed -E 's/.* ([0-9]+\\.[0-9]+\\.[0-9]+).*/\\1/')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    mkdir ${prefix}__recon_all

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        freesurfer: \$(mri_convert -version | grep "freesurfer" | sed -E 's/.* ([0-9]+\\.[0-9]+\\.[0-9]+).*/\\1/')
    END_VERSIONS

    function handle_code () {
    local code=\$?
    ignore=( 1 )
    exit \$([[ " \${ignore[@]} " =~ " \$code " ]] && echo 0 || echo \$code)
    }
    trap 'handle_code' ERR

    recon-all --help
    """
}
