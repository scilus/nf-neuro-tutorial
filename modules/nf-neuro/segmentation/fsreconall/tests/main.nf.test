nextflow_process {

    name "Test Process SEGMENTATION_FSRECONALL"
    script "../main.nf"
    process "SEGMENTATION_FSRECONALL"

    tag "modules"
    tag "modules_nfcore"
    tag "segmentation"
    tag "segmentation/fsreconall"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "freesurfer.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("segmentation - fsreconall") {

        when {
            config "./nextflow_debug.config"

            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/anat_image.nii.gz"),
                        file("\${test_data_directory}/license.txt")
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.recon_all_out_folder.get(0).get(1)).name,
                    process.out.versions
                ).match() }
            )
        }

    }


    test("segmentation - fsreconall - nolicense") {

        when {
            // Not in mode debug + no license = should fail.
            config "./nextflow.config"

            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/anat_image.nii.gz"),
                        []
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert ! process.success }
            )
        }

    }

    test("segmentation - fsreconall - stub-run") {
        options "-stub-run"
        when {
            config "./nextflow_debug.config"

            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/anat_image.nii.gz"),
                        file("\${test_data_directory}/license.txt")
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }

    }

}
