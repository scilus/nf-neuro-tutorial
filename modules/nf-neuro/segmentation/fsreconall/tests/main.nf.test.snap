{
    "segmentation - fsreconall - stub-run": {
        "content": [
            [
                "versions.yml:md5,2a7185588fa8c664d53c4c08637285e6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.1"
        },
        "timestamp": "2024-12-12T13:55:39.259005"
    },
    "segmentation - fsreconall": {
        "content": [
            "test__recon_all",
            [
                "versions.yml:md5,2a7185588fa8c664d53c4c08637285e6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-11-22T19:24:52.417991457"
    }
}